boskos-acquire/0.1/boskos-acquire.yaml

```diff
@@ -0,0 +1,67 @@
+apiVersion: tekton.dev/v1beta1
+kind: Task
+metadata:
+  name: boskos-acquire
+  labels:
+    app.kubernetes.io/version: "0.1"
+  annotations:
+    tekton.dev/pipelines.minVersion: "0.12.1"
+    tekton.dev/tags: "boskos,test"
+spec:
+  description: |
+    The boskos-acquire Task will request a resource of the specified type from the
+    server-url. If successful, it will start a pod that will run the boskosctl heartbeat
+    command. When you are done with the resource, release it with boskos-release.
+  params:
+  - name: server-url
+    description: The URL of the running boskos server
+    default: "http://boskos.test-pods.svc.cluster.local"
+  - name: type
+    description: |
+      The type of resource to request. Resource types are specified in the resource
+      ConfigMap provided to the Boskos server.
+    default: gke-project
+  - name: owner-name
+    description: A string that identifies the owner of the leased resource to request.
+  results:
+  - name: leased-resource
+    description: The name of the leased resource
+  steps:
+  - name: boskosctl-acquire
+    image: gcr.io/k8s-staging-boskos/boskosctl@sha256:a7fc984732c5dd0b4e0fe0a92e2730fa4b6bddecd0f6f6c7c6b5501abe4ab105
+    script: |
+      RESOURCE=$(boskosctl acquire \
+        --server-url=$(params.server-url) \
+        --owner-name=$(params.owner-name) \
+        --type=$(params.type) \
+        --state=free \
+        --target-state=busy)
+      echo $RESOURCE > /workspace/full-resource-output.json
+      echo $RESOURCE | jq -rj ".name" > /tekton/results/leased-resource
+  - name: create-boskosctl-heartbeat-pod
+    image: ubuntu
```

Comment:
I was gonna comment that same. We can combined the two step into one indeed :angel:

Reply:
i got really excited about this and then i realized that we'd need a way to run these lines:

```
      FULL_RESOURCE_OUTPUT=$(cat /workspace/full-resource-output.json)
      LEASED_RESOURCE=$(cat /tekton/results/leased-resource)
```

I guess if we're running cat in the next step, we'd be able to run cat inline?

alternatively: what about using a smaller image here instead of ubuntu, e.g. busy box?

Reply:
@bobcatfish I don't get what does it change ? writing the file vs piping it to `kubectl` is doing the same. Instead of writing `/workspace/heartbeat.yaml` and then running kubectl with it, we *pipe* what was written in that file directly to `kubectl`, nothing else changes :thinking:

Reply:
My train of thought was something like:
1. cool we can just use the kubectl image!
2. oh wait, what about those values from the files (not in andrea's example)
3. oh we'll need to run a shell
4. oh we're already running a shell
5. i guess the advantage here is that we're running 1 image instead of 2?

I think it's around (3): now we need a shell in the image that runs kubectl and (5) that i get a lot less excited.

it seems clearer to me to have 2 steps, one that makes the file and another that uses the file vs squishing them together for efficiency reasons. so im wondering if using smaller images for both might be another solution?

the kubectl image here already only contains sh and does not contain bash:

```
(⎈ |euca:default)➜  catalog git:(boskos) ✗ docker run -it --entrypoint /bin/bash lachlanevenson/k8s-kubectl
docker: Error response from daemon: OCI runtime create failed: container_linux.go:346: starting container process caused "exec: \"/bin/bash\": stat /bin/bash: no such file or directory": unknown.
(⎈ |euca:default)➜  catalog git:(boskos) ✗ docker run -it --entrypoint /bin/sh lachlanevenson/k8s-kubectl
~ #
```

Reply:
right, my question is why do we even need to writet the file (`heartbeat.yaml`). On bash/sh, I don't think there is any "bash" specifics in the script used, it would work with `/bin/sh` just as well, so using `lachlanevenson/k8s-kubectl` works just fine

Reply:
note that @a-roberts  just added a kubectl task (https://github.com/tektoncd/catalog/pull/394) this is getting me thinking about sharing steps again (https://github.com/tektoncd/pipeline/issues/1260) 🤔 

> right, my question is why do we even need to writet the file (heartbeat.yaml)

we dont need to write it, i think the trade off is something like:

1. is it clearer when it's 2 steps (maybe a little? maybe not? personally i think 2 steps is clearer than a script)
2. does the image contain everything you need to make it 1 step (it contains /bin/sh but not bash, so we wouldnt have the full power of bash replacement)

I think it's a bit clearer with 2 steps, so if i can switch out ubuntu for a faster image, does that address your concerns, or do you feel like the overhead of writing the file before applying it isn't worth it?

(oh oh oh - if only you could pipe the stdout of one step into another 😍😍😍  https://github.com/tektoncd/pipeline/issues/2925 ??? what do you think? there'd be more file i/o for sure, but wouldn't it be so clear ???)

Reply:
I would definitely use a smaller image (busybox?).
 
Personally I would just collapse the two steps, not only because it is slightly more efficient, but also because it would seem more readable to me: 1 step gets the lease from boskos, 1 step starts the heartbeat. 
If you prefer having the two separate steps I don't mind.

How would you implement piping stdout from one step into another? Would the two steps then run in parallel?  If not I feel it would be misleading to talk about "pipe".

Reply:
> Would the two steps then run in parallel? If not I feel it would be misleading to talk about "pipe".

whoa TIL that all the processes connected with pipes are started together (https://en.wikipedia.org/wiki/Pipeline_(Unix)#Implementation) - i had no idea! i thought `process1 | process2 | process3` would run process1, then 2, then 3, i didnt realize theyd all start together

TBH that's even more similar to the way our steps work, since they do actually all start simultaneously. to do this in a truly unix pipe-like manner, we could implement "inter process pipes" if we wanted to

but what im suggesting in #2925 is much simpler, not actually streaming data between steps (maybe day? probably crazy) but just providing stdout to the next step once the previous step is completed

Reply:
> Personally I would just collapse the two steps, not only because it is slightly more efficient, but also because it would seem more readable to me: 1 step gets the lease from boskos, 1 step starts the heartbeat.
If you prefer having the two separate steps I don't mind.

we're probably just getting into personal preference now but i'll say my final 2 cents - having 2 steps makes sense to me b/c:

1. The step that creates the config needs a shell, so use an image with a shell
2. The step that applies the config doesn't need a shell, it just needs kubectl, so don't use a shell, just use kubectl

Reply:
too bad the kubectl image is 50mb X'D (maybe they should remove sh XD)

Reply:
i bet kubectl needs sh to run

Reply:
> > Personally I would just collapse the two steps, not only because it is slightly more efficient, but also because it would seem more readable to me: 1 step gets the lease from boskos, 1 step starts the heartbeat.
> > If you prefer having the two separate steps I don't mind.
> 
> we're probably just getting into personal preference now but i'll say my final 2 cents - having 2 steps makes sense to me b/c:
> 
>     1. The step that creates the config needs a shell, so use an image with a shell
>     2. The step that applies the config doesn't need a shell, it just needs kubectl, so don't use a shell, just use kubectl

I am not actually sure it's too much about personal preferences, but also about user experiences. More step means more containers (if you have things like LimitRange, …, this means more resource reserved), more different images means more images to pull, more image to store on nodes :upside_down_face:.

I think our point (andrea and I) is that you don't need to "create" the config (as a file), you can just pipe it to kubectl. And as you said, the `kubectl` image already have `sh`, so it's all you need.

Any way, not wanting to block the PR for that :wink:

Reply:
> but also about user experiences. More step means more containers (if you have things like LimitRange, …, this means more resource reserved), more different images means more images to pull, more image to store on nodes 

I guess, feels like a strange place to be optimizing to me: I also have a dream of having many tiny images which do one thing well and only that thing, which can be cleanly composed together.

anyway maybe we can revisit it as we go - seems like it's not too bad to have the 2 images instead of one for now, so seems like it's okay to leave it for now

Reply:
I tried making my own image for this. However, the only way I was able to getting to work was by making a curl request to the k8s cluster directly. If the k8s api changes than it would have to be updated every time. I tried finding a k8s-cli slim but my google-foo failed me :sob:. It would be cool to have a container that had a bunch of kubectl slim that just basic verb/object capabilities. I might even do it as a side-project, tbh.